Structured declaration & assignment fails

**Describe the bug**

This example from [language docs](https://github.com/microsoft/BosqueLanguage/blob/master/docs/language/overview.md#64-structured-declaration-and-assignment) can't be compiled.

```js
namespace NSMain;

entrypoint function main(): Int {
  {f=let x, g=@[let y, let z]} = {f=1, g=@[2, 3]};
  return y;
}
```

It fails with: `Parse error -- ["declarations.bsq",6,"Literal match is not allowed"]`

@mrkmarron 

I'm getting this C++ compiler error:

```shell
Error -- Error: Command failed: clang++ -g -DBDEBUG -std=c++17 -o declarations /Users/brakmic/projects/BosqueLanguage/impl/bin/scratch/cpp/*.cpp
/Users/brakmic/projects/BosqueLanguage/impl/bin/scratch/cpp/main.cpp:37:18: warning: ISO C++ does not allow indirection on operand of type 'BSQ::Value' (aka 'void *') [-Wvoid-ptr-dereference]
    _tmp_6I36I = *_tmp_0I33I.atFixed<MIRPropertyEnum::g>();
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/brakmic/projects/BosqueLanguage/impl/bin/scratch/cpp/main.cpp:37:16: error: no viable overloaded '='
    _tmp_6I36I = *_tmp_0I33I.atFixed<MIRPropertyEnum::g>();
    ~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/brakmic/projects/BosqueLanguage/impl/bin/scratch/cpp/bsqcustom/../bsqvalue.h:746:15: note: candidate function not viable: cannot convert argument of incomplete type 'void' to 'const BSQ::BSQTuple' for 1st argument
    BSQTuple& operator=(const BSQTuple& src)
              ^
/Users/brakmic/projects/BosqueLanguage/impl/bin/scratch/cpp/bsqcustom/../bsqvalue.h:759:15: note: candidate function not viable: cannot convert argument of incomplete type 'void' to 'BSQ::BSQTuple' for 1st argument
    BSQTuple& operator=(BSQTuple&& src)
              ^
/Users/brakmic/projects/BosqueLanguage/impl/bin/scratch/cpp/main.cpp:40:18: warning: ISO C++ does not allow indirection on operand of type 'BSQ::Value' (aka 'void *') [-Wvoid-ptr-dereference]
    _tmp_8I39I = *_tmp_0I33I.atFixed<MIRPropertyEnum::g>();
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/brakmic/projects/BosqueLanguage/impl/bin/scratch/cpp/main.cpp:40:16: error: no viable overloaded '='
    _tmp_8I39I = *_tmp_0I33I.atFixed<MIRPropertyEnum::g>();
    ~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/brakmic/projects/BosqueLanguage/impl/bin/scratch/cpp/bsqcustom/../bsqvalue.h:746:15: note: candidate function not viable: cannot convert argument of incomplete type 'void' to 'const BSQ::BSQTuple' for 1st argument
    BSQTuple& operator=(const BSQTuple& src)
              ^
/Users/brakmic/projects/BosqueLanguage/impl/bin/scratch/cpp/bsqcustom/../bsqvalue.h:759:15: note: candidate function not viable: cannot convert argument of incomplete type 'void' to 'BSQ::BSQTuple' for 1st argument
    BSQTuple& operator=(BSQTuple&& src)
              ^
2 warnings and 2 errors generated.
```

--- a/impl/src/ast/parser.ts
+++ b/impl/src/ast/parser.ts
@@ -1792,11 +1792,11 @@ class Parser {
         }
     }
 
-    private parseMapEntryConstructorExpression(): Expression {
+    private parseMapEntryConstructorExpression(disallowMapEntry?: boolean): Expression {
         const sinfo = this.getCurrentSrcInfo();
         const exp = this.parseOrExpression();
 
-        if (this.testAndConsumeTokenIf("=>")) {
+        if (!(disallowMapEntry) && this.testAndConsumeTokenIf("=>")) {
             return new MapEntryConstructorExpression(sinfo, exp, this.parseMapEntryConstructorExpression());
         }
         else {
@@ -1804,14 +1804,14 @@ class Parser {
         }
     }
 
-    private parseSelectExpression(): Expression {
+    private parseSelectExpression(disallowMapEntry?: boolean): Expression {
         const sinfo = this.getCurrentSrcInfo();
-        const texp = this.parseMapEntryConstructorExpression();
+        const texp = this.parseMapEntryConstructorExpression(disallowMapEntry);
 
         if (this.testAndConsumeTokenIf("?")) {
-            const exp1 = this.parseMapEntryConstructorExpression();
+            const exp1 = this.parseMapEntryConstructorExpression(disallowMapEntry);
             this.ensureAndConsumeToken(":");
-            const exp2 = this.parseSelectExpression();
+            const exp2 = this.parseSelectExpression(disallowMapEntry);
 
             return new SelectExpression(sinfo, texp, exp1, exp2);
         }
@@ -2482,7 +2482,7 @@ class Parser {
 
         let whencheck = undefined;
         if (this.testAndConsumeTokenIf("when")) {
-            whencheck = this.parseExpression();
+            whencheck = this.parseSelectExpression(true);
         }
 
         if (istypematch) {

--- a/impl/src/tooling/aot/cpptype_emitter.ts
+++ b/impl/src/tooling/aot/cpptype_emitter.ts
@@ -472,7 +472,7 @@ class CPPTypeEmitter {
                 }
                 else {
                     if(trinto.base === "BSQTuple" || trinto.base === "BSQRecord") {
-                        return `*${exp}`
+                        return `*(BSQ_GET_VALUE_PTR(${exp}, ${trinto.base}))`;
                     }
                     else {
                         return `BSQ_GET_VALUE_PTR(${exp}, ${trinto.boxed})->bval`;
@@ -499,7 +499,7 @@ class CPPTypeEmitter {
                 }
                 else {
                     if(trinto.base === "BSQTuple" || trinto.base === "BSQRecord") {
-                        return `*${exp}`
+                        return `*(BSQ_GET_VALUE_PTR(${exp}, ${trinto.base}))`;
                     }
                     else {
                         return `BSQ_GET_VALUE_PTR(${exp}, ${trinto.boxed})->bval`;

--- a/impl/src/type_checker/type_checker.ts
+++ b/impl/src/type_checker/type_checker.ts
@@ -4240,7 +4240,9 @@ class TypeChecker {
                 opts = [tval, ntval];
             }
 
-            nexttype = this.m_assembly.restrictNotT(sexp, tmatch);
+            if(guard.optionalWhen === undefined) {
+                nexttype = this.m_assembly.restrictNotT(sexp, tmatch);
+            }
         }
         else {
             const sguard = guard as StructureMatchGuard;

